### ISSUE TITLE
layout managers needed

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/sun#48]
Preferably something based on springs and struts. Or Swing GridBagLayout.

To help with this sort of control-panel layout:

![screenshot_30](https://f.cloud.github.com/assets/3046552/2261992/692e56fc-9e55-11e3-9eb0-e3fe28b07982.png)

so we don't have to write this sort of code:

https://github.com/phetsims/energy-skate-park-basics/blob/master/js/view/EnergySkateParkBasicsControlPanel.js

**Comment by zepumph** [comment]
This may be getting some traction, marking for dev meeting to discuss the plan.

**Comment by samreid** [comment]
Some notes from https://github.com/phetsims/studio/issues/192#issuecomment-730570376

KP: Yes, changing language dynamically is highly desirable if we can accomplish it.

CM/JO: Layout managers are the right solution to this.
JO: We don't support setPreferredSize() or anything like that which 

JO: MiG layout is like a next generation layout manager.

KP: Before work begins on rosetta, there should be more discussion amongst the development team about options.  If there's a way to less-painfully bake in the smart layout.

KP: Current things in phet-io should be cleaned up before starting this.

SR: OK if it takes sim specific time for each sim?
KP: If we have support for that in scenery, so sims going forward support dynamic layout of text. There are reasons to switch languages within a sim.  That would resonate with our funders as well.  That will be for new sims, and we could revisit old sims as necessary.  Like with ES6, where we revisit a sim at our own pace without having to go 100% to the new pattern.

CM: I'm afraid we would choose the quickest path, which would put this responsibility in Node, which is a mistake.  And this is not just about Text.  We have needed layout managers since we started.  We wrestle with layout all the time.

KP: I recommend brainstorming what that would look like, a good long term solution.  What would you want to see?

CM: Just clarifying this is not a quick workaround for text, it helps with the overall layout problem.

JO: I'm passionate about this and would like to be involved.

SR: Yes, the common cases sound great, but we also need micro-layout managers, that can relayout() on demand. If each component has its own layout function which gets called at the right time.  There is another issue for this.

JO: Yes, I agree.

CM: I'd be excited to use this.

JO: I'd be excited to build this.

**Comment by zepumph** [comment]
Discussed during dev meeting today. @ariel-phet thinks we should bring this to quarterly planning. It would be good to bring it with a plan/brainstorm first.

We will come up with a sub-team  to take a first pass. Assigning to @jonathanolson, @pixelzoom, @samreid  to take a look first.

**Comment by pixelzoom** [comment]
Or first big decision is what

**Comment by pixelzoom** [comment]
Our first big decision is where layout responsibilities should live.   From https://en.wikipedia.org/wiki/Layout_manager

> Widget toolkits that provide this function can generally be classified into two groups:
> 
> * Those where the layout behavior is coded in special graphic containers. This is the case in XUL and the .NET Framework widget toolkit (both in Windows Forms and in XAML).
> * Those where the layout behavior is coded in layout managers, that can be applied to any graphic container. This is the case in the Swing widget toolkit that is part of the Java API.

Given PhET's history with LayoutBox, I'm guessing that we'll want to pursue the former approach, where layout is the responsibility for specific types of Nodes.

**Comment by pixelzoom** [comment]
Some useful links...

Visual Guide to Swing layout managers: https://docs.oracle.com/javase/tutorial/uiswing/layout/visual.html

Mig Layout: http://www.miglayout.com
* porting to other frameworks: http://www.migcalendar.com/miglayout/Porting.pdf
* cheat sheet: http://www.migcalendar.com/miglayout/mavensite/docs/cheatsheet.pdf

CSS layouts
* Grid: https://www.w3schools.com/css/css_grid.asp
* Flexbox: https://www.w3schools.com/css/css3_flexbox.asp

**Comment by samreid** [comment]
Some desired features and questions:

1. Automatic relayout when child components change size.
2. Could these sliders have the same track width?

![image](https://user-images.githubusercontent.com/679486/101415373-020b8200-38a5-11eb-8dd6-a6a64ba2af60.png)

3. Could we left-align radio buttons and text, but right-align the icons?  We do this in many (most?) cases, but we should make it easy to do.
![image](https://user-images.githubusercontent.com/679486/101415427-1ea7ba00-38a5-11eb-9f07-001388d279c6.png)

4. Automatically hide/remove separator if it has nothing to separate?

5. Align things in different control panels or different screens? (AlignBox helps solve this).  Will we ever want to line things up across sims?

6. Easy way to "indent" controls:

![image](https://user-images.githubusercontent.com/679486/101415674-91b13080-38a5-11eb-8170-9cc2334eaa94.png)

7. If the "none" label gets wider, should it make the entire panel wider, or make the slider track narrower?  Should this be determined by `maxWidth`?
![image](https://user-images.githubusercontent.com/679486/101415799-bf967500-38a5-11eb-859e-63d88b185c76.png)

8. Auto-collapse when elements are hidden by phet-io or by the code.

Also, how would we accomplish this:  A 200px wide panel with a slider1 (with long tickmark labels) in it, we want slider2 to have the same track width as slider1.  Bonus points if it is symmetrical, so that slider2 could be the one with longer tick labels.

It seems like a fundamental feature we would need for a layout component is `getMinimumDimension` and `setDimension`.  I'm not sure if this would be the same as `setBounds`.  http://www.miglayout.com/QuickStart.pdf points out that SWT only supports `preferredSize` while swing supports min/max/preferred.  The other building block we will need is for a child to be able to tell a parent when to relayout().  Can this always be done by just observing the bounds?

I can tell MiGLayout is highly revered, and it would be nice to have a documented standard for layout.  But (a) it looks nontrivial to port and (b) I'm not sure whether it does everything we would need and (c) I'm uncertain about the legibility of the API.  I wonder if we would be better off with our own implementation/API that does 90% of what we need with 20% of the effort?

**Comment by samreid** [comment]
Some interesting links:

Discussion about cassowary, which is the constraint based layout engine used in iOS
https://news.ycombinator.com/item?id=13125093

Autolayout
https://github.com/IjzerenHein/autolayout.js

![image](https://user-images.githubusercontent.com/679486/101709308-70d70f80-3a4c-11eb-85b4-ece865036a0e.png)

Apple's constraint API
https://developer.apple.com/documentation/quartzcore/caconstraintlayoutmanager

**Comment by samreid** [comment]
It seems like the conventional wisdom around constraint-based layouts is it can be very tricky to debug or understand, or if you put in a bad constraint then everything can go wrong easily.

We should also consider the cost to move pre-existing code forward to a new paradigm. I wonder if a sweet spot would be like the proposal in https://github.com/phetsims/joist/issues/608#issue-565036762, which allows us to use the same layout code we normally use--when object sizes change, the layout code runs again.

**Comment by samreid** [comment]
One of the issues that inspired this issue was https://github.com/phetsims/energy-forms-and-changes/issues/349, where labels are off-center if the text bounds change:

![image](https://user-images.githubusercontent.com/679486/101924512-06c68380-3b8e-11eb-88e9-a8e2f4ba82c2.png)

I'm trying to figure out how a "layout manager" would solve that problem, or how we would want that code to look after we have a layout manager.  The existing code with the problem is:

```js
    labelNode.translation = new Vector2(
      beakerBounds.centerX - labelNode.bounds.width / 2,
      beakerBounds.maxY - beakerBounds.height * beaker.fluidProportionProperty.value + topEllipse.bounds.height * 1.1
    );
```

If this is changed to:

```js
labelNode.y = beakerBounds.maxY - beakerBounds.height * beaker.fluidProportionProperty.value + topEllipse.bounds.height * 1.1;
labelNode.boundsProperty.link( bounds => labelNode.setCenterX( beakerBounds.centerX ) );
```

Then the problem goes away:

![image](https://user-images.githubusercontent.com/679486/101924678-32e20480-3b8e-11eb-8d23-3e8f784de8c3.png)

Let's make sure our proposed solutions can address cases like this.

**Comment by samreid** [comment]
On slack, @jonathanolson said: 

> Is that something AlignBox solves right now?

> Just setting AlignBox's bounds to the region you want content positioned in, and then in this case just keeping the default of centering should work?

**Comment by samreid** [comment]
For comparison, here is the `link` code:

```js
labelNode.boundsProperty.link( () => labelNode.setCenter( bottomEllipse.bounds.center.plusXY( 0, -48 ) ) );
```

And here is the equivalent `AlignBox` implementation:

```js
const alignBox = new AlignBox( labelNode, {
  alignBounds: bottomEllipse.bounds.shiftedY( -48 )
} );
```

It's not clear to me which is preferable.

**Comment by pixelzoom** [comment]
I recall this issue being discussed during a project planning meeting, but I don't see any notes from that meeting here. My recollection is that @kathy-phet asked @jonathanolson if he could explore some of the ideas that he had in mind.  But I don't recall what was decided -- Is @jonathanolson going to do that? How much time? When will he bring ideas back to the subgroup?...

@kathy-phet @jonathanolson can you fill in some details here about what's next?

**Comment by kathy-phet** [comment]
Yes, Q1 goal is for @jonathanolson to spend up to 20 hours on exploration of approaches and share his findings/proposed directions with devs for feedback by end of March.

**Comment by jonathanolson** [comment]
I think it would be good to have a call with @samreid and @pixelzoom to show my prototype work and discuss. If that sounds good, let's schedule something!

Notably I've arrived at 3 main tools:

1. A way of expressing manual constraints which will automatically update, AND will handle coordinate transforms transparently, e.g.:

```js
  ManualConstraint.create( rootNode, [ nodeA, nodeB ], ( a, b ) => {
    b.left = a.right + 10;
    b.centerY = a.centerY;
  } );
```

Systems can be constructed with multiple manual constraints. Based on reading and formulating, I don't believe it's an effective use of time to create a linear-equation-style constraint satisfaction system like "Auto Layout", as it seems complicated and not user-friendly outside of GUIs.

2. An upgraded 1-dimensional layout container, that could potentially be a drop-in replacement for LayoutBox. Notably it ALSO supports line-wrapping, line-spacing, justification options, custom per-cell margins/dimensions/alignment, and ways to grow to a "preferred" size in a user-controlled manner. It supports basically everything that can be done with flexbox.

3. A grid layout container, that should support the majority of what can be done with gridbag or mig (again, with custom per-cell margins/dimensions/alignment, cells that span multiple rows/columns, etc.)

Notably for (2) and (3), they also support setting preferred width/height on cell contents, so that you can mark some cells as being able to expand to fit available room. They themselves both support preferred sizes, so you could do the equivalent of "hey VBox or grid, take up these exact dimensions, and fill it THIS way"

Furthermore, (2) and (3) work nicely by adding a Node field: `layoutOptions`. The idiomatic way would be to provide any layout options for parent containers in the Node itself, so no intermediate Nodes are required (should have less overhead, and be much simpler).

Additionally, (2) and (3) are implemented with full generality, so while I have FlowBox and GridBox as useful abstractions, you can also manually create FlowConstraints or GridConstraints for laying out things that aren't able to have a common parent node.

I feel like the above hits ALL of my desired goals for layout, and seems to be running smoothly, but I'd really like feedback on things I might be blind to.

**Comment by samreid** [comment]
A few questions based on a quick read-through:

1. What will happen if you create a cycle using `ManualConstraint`?  Stack overflow?
2. What are the relationships between (1), (2) and (3).  Is one of them implemented using the other?  Are they all independent?
3. Should we ask how much the memory footprint increases by adding `layoutOptionsChangedEmitter` and `layoutOptions` to every Node?  I recall @zepumph and I did analysis like this when converting things to TinyEmitter and I recall it was on the low side.
4. Aside from a meeting/discussion and what comes out of that, what other steps do you envision before trying this out in production simulation code?  Off the top of my head, I considered: code review, having someone take it for a test drive, merging to master.

**Comment by jonathanolson** [comment]
> What will happen if you create a cycle using ManualConstraint? Stack overflow?

Yes pretty much. Auto Layout does handle "trying to satisfy multiple constraints as best as possible", but I've read it seems very hard to predict/debug.

> What are the relationships between (1), (2) and (3)? Is one of them implemented using the other? Are they all independent?

Independent.

> Should we ask how much the memory footprint increases by adding layoutOptionsChangedEmitter and layoutOptions to every Node?

I don't mind looking, but I believe it's likely to be not much and worth it to me.

> Aside from a meeting/discussion and what comes out of that, what other steps do you envision before trying this out in production simulation code? Off the top of my head, I considered: code review, having someone take it for a test drive, merging to master.

It's still fairly prototyped. I'd ideally like to see if the API/design should have changes (and I have a few pointed questions) before refining the code. Once that's settled/changed, code review, test-drive, merges, etc. sound good (in whatever order works best). Presumably lots of documentation and examples would help (I'll probably create some examples in advance of the discussion).

**Comment by pixelzoom** [comment]
On Slack, @jonathanolson said (presumably about the above commits):

> I've put up some examples of the current API (scenery checkout branch "layout", `grunt` to build the local copy, then scenery/doc/layout.html) if either of you wants to take a peek (note that dragging the bottom-right corner of selected examples controls the preferred width/height of the content)

**Comment by pixelzoom** [comment]
2/23/21 Zoom meeting: @jonathanolson @samreid @pixelzoom 

We reviewed @jonathanolson's prototype, and it's looking really excellent.  He kept notes about some suggestons, and will continue to work on it.  When he feels like it's ready, he'll put things in a new scenery/js/layout/ directory, merge to master, and PSA to the dev team. 

Still to discuss:
* When/where to start using this in production code? Should we approach this like bamboo?
* What to do about existing layout classes (LayoutBox, VBox, HBox, AlignBox,...)?  Deprecate? Re-implement using new layout managers?
* How/when to add resize support to existing PhET UI components - which components, priority, backward compatibility, etc.

**Comment by pixelzoom** [comment]
In Slack, @samreid said:

> Some other questions I wrote down that we didn’t get to today:
Get away from memoize/mixin/trait.
Put preferred size in every node?  Or use a lazy instantiation approach?
We have no schema for checking layoutOptions, right?
Is that a problem?
Could FlowLayout be implemented using GridLayout?
Does ManualConstraint supercede AlignBox?
What should be used instead of VBox/HBox?

**Comment by jonathanolson** [comment]
> Get away from memoize/mixin/trait.
> Put preferred size in every node? Or use a lazy instantiation approach?

These seem like more issues in our IDE, maybe we could fix those? Should ParallelDOM and Paintable and Imageable all also be put in Node? (It would presumably blow up our memory usage, slow things down, and cause maintenance headaches).

> We have no schema for checking layoutOptions, right? Is that a problem?

It sounds important to be able to check this.

> Could FlowLayout be implemented using GridLayout?

No, they are fundamentally different.

> Does ManualConstraint supercede AlignBox?

No, ManualConstraint doesn't adjust localBounds. I'd presumably add something like AlignBox.

> What should be used instead of VBox/HBox?

I'm planning on looking into whether HBox/VBox can be changed to be backed by FlowBox.

**Comment by pixelzoom** [comment]
This issue hasn't moved since in almost 6 weeks.  Where are we at?

**Comment by pixelzoom** [comment]
Here's where we're at...

We need to discuss next steps listed in: https://github.com/phetsims/sun/issues/48#issuecomment-784511849

We need to determine if @samreid's questions have been answered/addressed. @jonathanolson answered those questions in https://github.com/phetsims/sun/issues/48#issuecomment-785399207. @samreid does this address your questions, or do they require additional discussion?

Finally, we need to schedule a subgroup meeting for discussion.

**Comment by samreid** [comment]
Those questions have been answered, with some more conversation about trait/mixin continuing in https://github.com/phetsims/phet-info/issues/143#issuecomment-792919089.

Also, would be good to touch base on simple layout problems such as https://github.com/phetsims/sun/issues/48#issuecomment-743275274 now that the layout engine is further along.  For cases like that would we still just manually listen/layout?

**Comment by pixelzoom** [comment]
This issue has stalled out. The last time this subgroup met was Feb 23.  Since @samreid has unassigned, I guess I'll do that same, and await further instructions from @jonathanolson or @kathy-phet.

**Comment by samreid** [comment]
The commit from yesterday introduced this code to, say, Panel:

```js
class Panel extends WidthSizable( HeightSizable( Node ) ) { ... }
```

This is causing TypeScript compilation errors to appear on CT, like:

```
../../../greenhouse-effect/js/common/view/LandscapeObservationWindow.ts:240:34 - error TS2345: Argument of type 'EnergyBalancePanel' is not assignable to parameter of type 'Node'.
  Type 'EnergyBalancePanel' is missing the following properties from type 'Node': _id, _instances, _rootedDisplays, _drawables, and 431 more.

240     this.controlsLayer.addChild( energyBalancePanel );
```

**Comment by pixelzoom** [comment]
This may be an unpopular opinion.  But I think we should avoid introducing more mixins/traits into common code until we have a pattern that works well for TS.   

In the meantime:
* If the mixin is something like `class Image extends Imageable( Node )`, then you can workaround it via `unknown as Node` - ugly but effective.  
* If the mixin adds fields or methods (like `ParallelDOM`), there is no workaround, and you have to resort to `@ts-ignore`.

**Comment by jonathanolson** [comment]
There are also ways of creating d.ts files (like done for Panel) that get things to work nicely. However this does duplicate some of the API of things (seems reasonable if this is short-term).

**Comment by jonathanolson** [comment]
https://github.com/phetsims/sun/commit/da619e1be17d20cd64cafdf8af36c4ac5c72cb4b refits Panel to use LayoutConstraint internally, so we don't need to duplicate listener/locking logic. It also uses a better locking approach (which will work with changing minimum dimensions of a sizable node causing reentrance).

@samreid or @pixelzoom, this could be a decent commit to review.

It passed aqua checks and snapshot comparison.

**Comment by pixelzoom** [comment]
@jonathanolson What is "locking" in this context?

**Comment by jonathanolson** [comment]
Locking refers to the prevention of certain recursion/reentrance in layout that would result in infinite loops.

Panel had:
```js
    // flag for preventing recursion
    let backgroundUpdateInProgress = false;

///////

      // Check that an update isn't already in progress, lest we end up with some nasty recursion.  For details, please
      // see https://github.com/phetsims/sun/issues/110 and https://github.com/phetsims/molecule-shapes/issues/130.
      if ( backgroundUpdateInProgress ) {
        return;
      }
```

LayoutConstraint has an improved version of this, that will prevent reentrant layouts, BUT will record whether one was attempted, and will re-run the layout afterward.

This is needed, since the Panel layout locking approach left layout in "stale" approaches - e.g. if the panel changes the preferred size of its contents during layout, it could trigger a resize on the size of the contents (which would then not be respected, since it would trigger a reentrant layout request).

**Comment by samreid** [comment]
I read through Panel.js and LayoutConstraint.js.  It seems reasonable and I didn't have many questions.  Do you imagine we would generalize PanelConstraint in the future? Can you recommend how to exercise this code for testing?

**Comment by pixelzoom** [comment]
Review of LayoutConstraint and Panel...

LayoutConstraint.js:

- [ ] The doc for `_layoutLockCount` is incorrect.  It's not a boolean, it's a count. `layout` runs only if the count is 0.

```js
    // @private {number} - Prevents layout() from running while true. Generally will be unlocked and laid out.
    this._layoutLockCount = 0;
```

- [ ] `layout` is protected because it's abstract and the subclass needs to define it.   But that opens up the possibility of the subclass bypassing the logic in `updateLayout`, ignoring `_layoutLockCount`, and calling `layout` directly.  Add some documentation indicating that only `updateLayout` should be calling `layout`. 

Panel.js:

- [ ] You're passing `options` directly to PanelConstraint. The nest options pattern should be used here.

- [ ] You haven't provided defaults for any of the options in PanelConstraint. And based on the assertions in PanelConstraint, it looks like all of the options are actually required. So `options` should be `config`, and you should be using `required` to verify that they are provided.

**Comment by pixelzoom** [comment]
From our last subteam disssion on 2/23/21 (!!):

> Still to discuss:
> * When/where to start using this in production code? Should we approach this like bamboo, and use initially in a sim or two, so that it's easier to make refinements?
> * This is very powerful, complicated, likely a relatively steep learning curve. What doc, tools, etc. should be in place to support developers?
> * How/when to add resize support to existing PhET UI components - which components, priority, cost, backward compatibility, etc.
> * What to do about existing layout classes? (LayoutBox, VBox, HBox, AlignBox,...) Deprecate? Re-implement using new layout managers?
> * Do any aspects of layout managers need to be exposed in the PhET-iO API?
> * Does this affect any aspects of a11y, such as traversal?

@jonathanolson @samreid When can we discuss? And when do we expect to be using this in sims?

**Comment by jonathanolson** [comment]
> You're passing options directly to PanelConstraint. The nest options pattern should be used here.

Are you proposing we change the Panel API like so? (strong preference not to do so):

```js
new Panel( content, {
  fill: 'white',
  panelConstraintOptions: {
    xMargin: 5,
    lineWidth: 2
  }
} );
```

> You haven't provided defaults for any of the options in PanelConstraint. And based on the assertions in PanelConstraint, it looks like all of the options are actually required. So options should be config, and you should be using required to verify that they are provided.

Default values are provided in the Panel optionize. Now that it's in TypeScript, it's typing nicely, and we are guaranteed these values.

**Comment by jonathanolson** [comment]
> When/where to start using this in production code? Should we approach this like bamboo, and use initially in a sim or two, so that it's easier to make refinements?

I believe soon for general purpose. I've been using these in my simulations, and I'm pretty confident in the FlowBox API (and associated places). I'm slightly less confident in GridBox (want the time to examine the interaction of alignments/stretch/grow). ManualConstraint I believe is also ready. Checkbox/Panel already includes things based on LayoutConstraint. I've definitely refined FlowBox a moderate amount since introduction. If we can have HBox/VBox backed by FlowBox (which is my goal), then it would be a quick switch over (I'd want unit tests and things first).

> This is very powerful, complicated, likely a relatively steep learning curve. What doc, tools, etc. should be in place to support developers?

I have documentation under scenery/doc/layout.html, and a layout screen in sun's demo. I'd like to expand on all of this to flesh out examples of common layouts, and detail all of the features. I have a to-do item that is an overlay that would show information about all of the layout containers in a sim in order to debug things.

> How/when to add resize support to existing PhET UI components - which components, priority, cost, backward compatibility, etc.

Added to Panel/Checkbox/AlignBox. Planned for ComboBox, Slider and buttons. Thoughts on other components?

> What to do about existing layout classes? (LayoutBox, VBox, HBox, AlignBox,...) Deprecate? Re-implement using new layout managers?

I've attempted to make FlowBox completely compatible with VBox/HBox. My plan is to swap the implementation of those to FlowBox (with the requisite orientation), and presumably LayoutBox for FlowBox. AlignBox has its own benefits (is now set up to be resizable, and has layout benefits and margins that are needed in general) so it should be kept.

> Do any aspects of layout managers need to be exposed in the PhET-iO API?

Not that I'm aware of, but those things could possibly be added.

> Does this affect any aspects of a11y, such as traversal?

No. It's just like the current containers for a11y.

**Comment by jonathanolson** [comment]
Executed the HBox/VBox switch-over in the commit above. All automated testing is looking good.

**Comment by jonathanolson** [comment]
I'll have it ready for a full code review quite soon.

**Comment by jonathanolson** [comment]
I believe the layout work is ready for an initial code review. I have one TODO listed in GridConstraint (will handle), and a few enhancements, but the bulk of it is ready (and it's backing HBox/VBox now for Flow layout). I'll also be moving over more components to be resizable (notably ComboBox/Slider/Rectangular buttons/AccordionBox/NumberDisplay/RadioButton).

1. build scenery
2. read scenery/docs/layout.html (it describes a lot, and shows examples)
3. view some of the demos in the sun demo 4th screen (for layout) - contained in LayoutScreenView
4. review all of scenery/layout (with the exception of LayoutBox)
5. review Panel and Checkbox's changes (they were made resizable, and layout was moved to separate LayoutConstraints that have many advantages)

**Comment by jonathanolson** [comment]
Updated sun rectangular buttons with resizability above.

**Comment by pixelzoom** [comment]
Since this is a long-ish issue, and still in sun (instead of scenery) .... I'm going to do the code review in a new scenery issue, https://github.com/phetsims/scenery/issues/1419.
